### PR TITLE
Fix `exception_handlers` type hints

### LINKED ISSUE #1142: mypy complains about `exception_handlers` dict with `HTTPException` key
### Checklist

- [x] The bug is reproducible against the latest release and/or `master`.
- [x] There are no similar issues or pull requests to fix it yet.

Related, but not the same:
* #1034 
* #1139 
* #949

### To reproduce

Given an example.py: 

```python
from starlette.applications import Starlette
from starlette.exceptions import HTTPException
from starlette.requests import Request
from starlette.responses import Response
from starlette.routing import Route

def handle_http_exception(request: Request, exc: HTTPException) -> Response:
    return Response('foo')


exception_handlers = {
    HTTPException: handle_http_exception
}
app = Starlette(exception_handlers=exception_handlers)
```

### Expected behavior

Mypy does not present an error.

### Actual behavior

```console
$ mypy example.py
example.py:14: error: Argument "exception_handlers" to "Starlette" has incompatible type "Dict[Type[HTTPException], Callable[[Request, HTTPException], Response]]"; expected "Optional[Dict[Union[int, Type[Exception]], Callable[..., Any]]]"
Found 1 error in 1 file (checked 1 source file)
```

### Debugging material

None

### Environment

- OS: Debian (`Linux ... 4.19.0-14-amd64 #1 SMP Debian 4.19.171-2 (2021-01-30) x86_64 GNU/Linux`)
- Python version: `Python 3.9.1`
- Starlette version: `starlette==0.14.2`

### Additional context

For propriety:

https://github.com/encode/starlette/blob/master/docs/exceptions.md#L42

https://github.com/encode/starlette/blob/ed73b97c9b8f4aa527eed88032c04ce9fe109a3c/starlette/applications.py#L44

I think this is a Starlette issue and not a mypy issue; apologies if I'm wrong here.

### PR DESCRIPTION
**Pull request by aminalaee**
Closes #1142 and replaces #1139.

Changes `exception_handlers` type hints:

```python
typing.Mapping[
    typing.Any, typing.Callable[[Request, Exception], Response]
]
```

I also changed the callable type hints, I think it makes sense to be more explicit, this will catch user-defined callable mismatches (?).

As mentioned in #1139 we can't use `typing.Mapping[typing.Union[int, ..], ...]` since dict keys are not covariant.<del>I think we could use `typing.TypedDict` but only after 3.8</del>.

### CODE REVIEW

File: starlette/exceptions.py [hunk 1]
```diff
@@ -22,7 +22,12 @@ def __repr__(self) -> str:
 
 class ExceptionMiddleware:
     def __init__(
-        self, app: ASGIApp, handlers: dict = None, debug: bool = False
+        self,
+        app: ASGIApp,
+        handlers: typing.Mapping[
+            typing.Any, typing.Callable[[Request, Exception], Response]
+        ] = None,
```
Comment (k4black) [inline starlette/exceptions.py:29]: It's quite a complicated type, should we move it to `starlette.types`? Like `ExceptionHandler` type

**Comment by k4black** [comment]
Speaking about `typing.TypedDict` you can add python version check as a nice way to show that you'll migrate to dict when 3.7 will be deprecated

File: starlette/exceptions.py (hunk 1 above)
Comment (Kludex) [reply starlette/exceptions.py:29]: The current types on `types.py` are going to be removed at some point: https://github.com/encode/starlette/issues/1217

That file will only contain this type at some point. Is that a problem? :thinking:

Comment (k4black) [reply starlette/exceptions.py:29]: Didn't notice the Issue,  
This mapping type just seems a bit complicated to me to be copied multiple times

Comment (aminalaee) [reply starlette/exceptions.py:29]: One drawback I noticed in VScode, haven't really checked and there might be some workarounds, is that they don't show function signatures when using the aliases as you mention. Not sure if it happens on PyCharm and other tools.

**Review by Kludex — APPROVED** [review]

**Comment by antonagestam** [comment]
@aminalaee @k4black FYI, I've opened a PR that improves on the typing added in this PR: https://github.com/encode/starlette/pull/1456

**Pull request merged**